lib/zip/file.rb

```diff
@@ -405,14 +405,13 @@ def check_file(path)
     def on_success_replace
       tmpfile      = get_tempfile
       tmp_filename = tmpfile.path
-      ObjectSpace.undefine_finalizer(tmpfile)
-      tmpfile.close
       if yield tmp_filename
-        ::File.rename(tmp_filename, name)
+        new_file = ::File.new(name, "w")
+        IO.copy_stream tmpfile, new_file
```

Comment:
The original code uses File.rename to move the tempfile into place, which is a wrapper around the C function rename(3) so it is fast and atomic.  The proposed change is undesirable because it is slow (reads/writes entire ZIP file again) but more importantly not atomic, so it can leave a corrupt ZIP file behind if it is interrupted.